crates/subspace-service/src/dsn/import_blocks/piece_validator.rs

```diff
@@ -0,0 +1,148 @@
+use async_trait::async_trait;
+use futures::StreamExt;
+use lru::LruCache;
+use parking_lot::Mutex;
+use std::error::Error;
+use std::num::NonZeroUsize;
+use subspace_archiving::archiver::is_piece_valid;
+use subspace_core_primitives::crypto::kzg::{test_public_parameters, Kzg};
+use subspace_core_primitives::{
+    Piece, PieceIndex, RecordsRoot, SegmentIndex, PIECES_IN_SEGMENT, RECORD_SIZE,
+};
+use subspace_networking::libp2p::PeerId;
+use subspace_networking::utils::piece_provider::PieceValidator;
+use subspace_networking::{Node, RootBlockRequest, RootBlockResponse};
+use tracing::{debug, error, trace, warn};
+
+const RECORDS_ROOTS_CACHE_SIZE: NonZeroUsize = NonZeroUsize::new(1_000_000).expect("Not zero; qed");
```

Comment:
When downloading root block chain we need to ensure the number of root blocks is not bigger than this number, otherwise node needs to crash because we'll not be able to verify pieces anyway.